modules/auxiliary/scanner/http/gitlab_graphql_user_enum.rb

```diff
@@ -0,0 +1,180 @@
+    ])
+  end
+
+  ##
+  # Send the GraphQL query to the /api/graphql endpoint. Despite being able to
+  # extract significantly more information, this request will only request
+  # usernames. The function will do some verification to ensure the received
+  # payload is the expected JSON.
+  #
+  # @param after [String] The parameter is used for paging because GitLab will only
+  #   return 100 results at a time. If no paging is needed this should be empty.
+  # @return [Hash] A Ruby Hash representation of the returned JSON data.
+  ##
+  def do_request(after)
+    graphql_query = '{"query": "query { users'
+    unless after.empty?
+      graphql_query += "(after:\\\"#{after}\\\")"
+    end
+    graphql_query.concat(' { pageInfo { hasNextPage, hasPreviousPage, endCursor, startCursor }, nodes { username } } }" }')
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/api/graphql'),
+      'ctype' => 'application/json',
+      'data' => graphql_query
+    })
+
+    fail_with(Failure::UnexpectedReply, "The target didn't respond with 200 OK") unless res&.code == 200
+    fail_with(Failure::UnexpectedReply, "The target didn't respond with an HTTP body") unless res.body
+
+    user_json = res.get_json_document
+    fail_with(Failure::UnexpectedReply, "The target didn't return a JSON body") if user_json.nil?
+
+    nodes = user_json.dig('data', 'users', 'nodes')
+    fail_with(Failure::UnexpectedReply, 'Could not find nodes in the JSON body') if nodes.nil?
+
+    user_json
+  end
+
+  ##
+  # Parses the JSON data returned by the server. Adds the usernames to
+  # the users array and adds them, indirectly, to create_credential_login.
+  # This function also determines if we need to request more data from
+  # the server.
+  #
+  # @param user_json [Hash] The JSON data provided by the server
+  # @param users [Array] An array to store new usernames in
+  # @return [String] An empty string or the "endCursor" to use with do_request
+  ##
+  def parse_json(user_json, users)
+    nodes = user_json.dig('data', 'users', 'nodes')
+    return '' if nodes.nil?
+
+    nodes.each do |node|
+      username = node['username']
+      store_username(username, node)
+      users.push(username)
+    end
+
+    query_paging_info = ''
+    more_data = user_json.dig('data', 'users', 'pageInfo', 'hasNextPage')
+    if !more_data.nil? && more_data == true
+      query_paging_info = user_json['data']['users']['pageInfo']['endCursor']
+    end
+
+    query_paging_info
+  end
+
+  def store_userlist(users, service)
+    loot = store_loot('gitlab.users', 'text/plain', rhost, users, nil, 'GitLab Users', service)
+    print_good("Userlist stored at #{loot}")
+  end
+
+  def store_username(username, json)
+    service = ssl ? 'https' : 'http'
+    service_data = {
+      address: rhost,
+      port: rport,
+      service_name: service,
+      protocol: 'tcp',
+      workspace_id: myworkspace_id,
+      proof: json
+    }
```

Comment:
There's the `#service_details` method provided by the HttpClient mixin which may be helpful here to reduce some of the code reuse https://github.com/rapid7/metasploit-framework/blob/9dc9546b443ea781bb8920665fc2dea0aa2c341f/lib/msf/core/exploit/remote/http_client.rb#L926. You should only need to provide the proof, everything else will be taken care of automatically.

Reply:
Good call. I switched to using `service_details` and followed the `create_credential_and_login` pattern a number of other modules use. e.g:

https://github.com/rapid7/metasploit-framework/blob/c491687b78ac56a5c23a315ec00089f6a08349bc/modules/auxiliary/scanner/http/jira_user_enum.rb#L77

Which cut back on a lot of code.